doc/source/ray-core/examples/rdt/grpo_contextual_bandits.py

```diff
@@ -0,0 +1,370 @@
+        [-diag, -diag],  # SW
+    ],
+    dtype=torch.float32,
+)
+
+TrajectorySlice = dict[str, torch.Tensor | int]
+
+
+class MLP(torch.nn.Sequential):  # Sized to ~50 MB of parameters.
+    def __init__(self):
+        layers = []
+        in_dim = STATE_DIM
+        for _ in range(50):
+            layers.append(torch.nn.Linear(in_dim, 512, bias=True))
+            layers.append(torch.nn.LayerNorm(512))
+            layers.append(torch.nn.ReLU())
+            layers.append(torch.nn.Dropout(0.1))
+            in_dim = 512
+
+        # Produce logits for each action.
+        layers.append(torch.nn.Linear(in_dim, ACTION_DIM, bias=True))
+        layers.append(torch.nn.Dropout(0.1))
+
+        super().__init__(*layers)
+
+
+def sample_unit_vector(dim: int = STATE_DIM, batch_size: int = 1) -> torch.Tensor:
+    """Sample unit vector(s) by normalizing Gaussian draws."""
+    v = torch.randn(batch_size, dim)
+    norms = v.norm(dim=-1, keepdim=True) + 1e-8
+    unit_vectors = v / norms  # [batch_size, STATE_DIM]
+
+    if batch_size == 1:
+        return unit_vectors.squeeze(0)
+    return unit_vectors
+
+
+@ray.remote
+class ReplayBuffer:
+    """Storage for scored trajectory slices."""
+
+    def __init__(self) -> None:
+        # Each entry stores (policy_version, TrajectorySlice with CPU tensors).
+        self.storage: list[tuple[int, TrajectorySlice]] = []
+        # The sum of policy versions; used for the weighted sampling calculation.
+        self.total = 0
+
+    def put(self, slice: TrajectorySlice) -> None:
+        self.storage.append((slice["policy_version"], slice))
+        self.total += slice["policy_version"]
+
+    def sample_from(self, n: int) -> list[TrajectorySlice]:
+        """Sample n scored trajectory slices."""
+        assert len(self.storage) > 0
+        # The probability of sampling a slice is proportional to its policy version.
+        probs = [version / self.total for version, _ in self.storage]
+        indices = list(range(len(self.storage)))
+        # Sample with replacement without exceeding the buffer's size.
+        n = min(n, len(self.storage))  
+        chosen = np.random.choice(indices, size=n, p=probs, replace=True)
+        return [self.storage[i][1] for i in chosen]
+
+    def size(self) -> int:
+        return len(self.storage)
+
+
+@ray.remote(num_gpus=1)
+class Scorer:
+    """Evaluates actions and assigns rewards to trajectory slices.
+
+    This scorer implements an analytic contextual bandit reward: for a 2D unit
+    context vector `s` and a discrete action `a` in {W, NW, N, NE, E, SE, S, SW},
+    reward is cosine_similarity(s, direction[a]) == dot(s, unit_direction[a]).
+    """
+
+    def __init__(self, replay_buffer) -> None:
+        self.replay_buffer = replay_buffer
+        self.action_dirs = ACTION_DIRECTIONS.to("cuda")  # [ACTION_DIM, STATE_DIM]
+
+    def enqueue_trajectory_batch(self, batched_slices: dict) -> None:
+        """Score a batched trajectory slice synchronously."""
```

Comment:
What does "synchronously" mean here?

Reply:
I asked GPT to convert the actor from async -> sync and I missed this docstring D:<